neqo-transport/src/cc/classic_cc.rs

```diff
@@ -281,12 +308,22 @@ impl<T: WindowAdjustment> CongestionControl for ClassicCongestionControl<T> {
             &[QlogMetric::BytesInFlight(self.bytes_in_flight)],
         );
 
+        // Lost PMTUD probes do not elicit a congestion control reaction.
+        let lost_packets: Vec<SentPacket> = lost_packets
+            .iter()
+            .filter(|pkt| !self.pmtud.is_pmtud_probe(pkt))
+            .cloned()
+            .collect();
+        if lost_packets.is_empty() {
```

Comment:
See also https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.peekable.

Reply:
@mxinden I need some Rust help to make this fly. I changed `detect_persistent_congestion` to be
```rust
    fn detect_persistent_congestion<'a, I>(
        &mut self,
        first_rtt_sample_time: Option<Instant>,
        prev_largest_acked_sent: Option<Instant>,
        pto: Duration,
        lost_packets: I,
    ) -> bool
    where
        I: Iterator<Item = &'a SentPacket>,
```
and have this in `on_packets_lost`:
```rust
        // Lost PMTUD probes do not elicit a congestion control reaction.
        let lost_packets = lost_packets.iter().filter(|pkt| !self.pmtud.is_probe(pkt));
        let Some(last_lost_packet) = lost_packets.last() else {
            return false;
        };

        let congestion = self.on_congestion_event(&last_lost_packet);
        let persistent_congestion = self.detect_persistent_congestion(
            first_rtt_sample_time,
            prev_largest_acked_sent,
            pto,
            lost_packets,
        );
```
But I get some borrow checker errors (`cannot borrow `*self` as mutable because it is also borrowed as immutable`) that I can't work through.

Reply:
If you are holding a mutable reference to `self.foo` and you call `self.bar()`, that's not going to work.  The problem here is that you are holding a reference to `self.pmtud` via the iterator. 

Options I can think of without sitting down with this:

* You can change `detect_persistent_congestion` to not take `&mut self` and instead take mutable (or non-mutable) references to the fields it cares about.
* You can move `self.pmtud` off (`std::mem::take`) before making the call, restoring it afterwards.
* You can wait for the rust language to be updated to be smarter about what members are touched in functions.

Reply:
* Just ask PMTUD for the packet size that is needed to identify a PMTUD probe and do this:

```rust
let probe_size = self.pmtud.min_probe_size();
let lost_packets = lost_packets.iter().filter(|pkt| pkt.len() < probe_size).rev().peekable();
let Some(last_lost_packet) = lost_packets.peek() {
  return false;
}

let congestion = self.on_congestion_event(&last_lost_packet);
        let persistent_congestion = self.detect_persistent_congestion(
            first_rtt_sample_time,
            prev_largest_acked_sent,
            pto,
            lost_packets.rev(), // flip it back over again
        );
```

Reply:
https://github.com/larseggert/neqo/pull/25

> You could instead define `self.pmtud.is_not_probe_fn()` to return a function like `Fn(pkt: SentPacket) -> bool` and pass that to filter instead. That maintains encapsulation, sort of, but I'm not sure that rust is that good about how it manages captured variables.

That is a neat idea @martinthomson, especially since it keeps the logic encapsulated within `pmtud.rs`.

In case you like it @larseggert, I implemented it in https://github.com/larseggert/neqo/pull/25.